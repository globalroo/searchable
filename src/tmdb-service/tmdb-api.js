import {
	getDiscoverMovieEndpoint,
	getMovieByIdEndpoint,
	getMovieCastByIdEndpoint,
	getMovieGenresEndpoint,
	getPersonByIdEndpoint,
	getPersonCreditsByIdEndpoint,
	getPopularMovieEndpoint,
	getRecommendedMoviesByIdEndpoint,
	getSearchMovieEndpoint,
	getSearchMultiEndpoint,
	getTrendingPersonByWeekEndpoint,
	getTrendingMoviesByDay,
	getTrendingMoviesByWeek,
	getTrendingTvByWeekEndpoint,
	getTvByIdEndpoint
} from "./tmdb-config";

export const MEDIA_TYPE = {
	TV: "tv",
	MOVIE: "movie",
	PERSON: "person"
};

export const defaultFailResponse = {
	page: 0,
	results: [],
	total_pages: 0,
	total_results: 0
};

const handleResponse = ({ response, payload }) => ({
	response,
	payload
});

/**
 * Naive safe response with capability to handle basic error state.
 * @async
 * @function safeFetchJson
 * @param {string} url - the endpoint to fetch
 * @param {object} options - fetch request options object default = {},
 * @param {object} failResponse - an expected empty response from tmdb default = defaultFailResponse
 * @return {object} { payload, response } - "payload" - holds JSON from the response or failResponse. "response" - holds the actual response from the server to handle error conditions
 */

export const safeFetchJson = async ({ url, options = {}, failResponse = defaultFailResponse }) => {
	let response, payload;
	try {
		response = await window.fetch(url, options);
		payload = await response.json();
	} catch (error) {
		response = { ok: false, statusText: `${error}` };
		payload = failResponse;
	}
	return handleResponse({ response, payload });
};

// API helpers

export const getPopularMovies = async () =>
	await safeFetchJson({
		url: getPopularMovieEndpoint()
	});

export const getSearchMovieTitleResults = async term =>
	await safeFetchJson({
		url: getSearchMovieEndpoint(term)
	});

export const getSearchMultiResults = async term =>
	await safeFetchJson({
		url: getSearchMultiEndpoint(term)
	});

export const getMovieById = async id =>
	await safeFetchJson({
		url: getMovieByIdEndpoint(id),
		failResponse: {}
	});

export const getRecommendedMoviesById = async id =>
	await safeFetchJson({
		url: getRecommendedMoviesByIdEndpoint(id)
	});

export const getMovieCastById = async id =>
	await safeFetchJson({
		url: getMovieCastByIdEndpoint(id)
	});

export const getPersonById = async id =>
	await safeFetchJson({
		url: getPersonByIdEndpoint(id),
		failResponse: {}
	});

export const getTvById = async id =>
	await safeFetchJson({
		url: getTvByIdEndpoint(id),
		failResponse: {}
	});

export const getDailyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByDay()
	});

export const getWeeklyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByWeek()
	});

export const getWeeklyTrendingPeople = async () =>
	await safeFetchJson({
		url: getTrendingPersonByWeekEndpoint()
	});

export const getWeeklyTrendingTv = async () =>
	await safeFetchJson({
		url: getTrendingTvByWeekEndpoint()
	});

export const getPersonCreditsById = async id =>
	await safeFetchJson({
		url: getPersonCreditsByIdEndpoint(id)
	});

export const getMovieGenres = async () =>
	await safeFetchJson({
		url: getMovieGenresEndpoint()
	});

export const getDiscoverMoviesByGenreId = async (id, query) => {
	const applyFilter = query
		? query
		: "&sort_by=popularity.desc&include_adult=false&vote_average.gte=6";
	return await safeFetchJson({
		url: `${getDiscoverMovieEndpoint()}${applyFilter}&with_genres=${id}`
	});
};
