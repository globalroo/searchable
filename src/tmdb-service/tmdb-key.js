let API_KEY = ""; //TODO: handle API key

export const getKey = () => {
	return API_KEY;
};
