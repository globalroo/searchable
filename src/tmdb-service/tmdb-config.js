import { getKey } from "./tmdb-key";

export const BASE_URL = "https://api.themoviedb.org/3";
export const BASE_IMG_PATH = "http://image.tmdb.org/t/p";

const API_KEY = getKey();

// Obtain image sizes here
// https://api.themoviedb.org/3/configuration?api_key=

export const API = {
	COMBINED_CREDITS: "/combined_credits",
	CREDITS: "/credits",
	DAY: "/day",
	DISCOVER_MOVIE: "/discover/movie",
	GENRES: "/genre/movie/list",
	MOVIE: "/movie",
	PERSON: "/person",
	POPULAR_MOVIES: "/movie/popular",
	RECOMMENDATIONS: "/recommendations",
	SEARCH_MOVIES: "/search/movie",
	SEARCH_MULTI: "/search/multi",
	SMALL_IMG_WIDTH: "/w300",
	LARGE_IMG_WIDTH: "/w1280",
	TRENDING: "/trending",
	TV: "/tv",
	WEEK: "/week"
};

export const getAuthorisedEndpoint = endpoint => `${BASE_URL}${endpoint}?api_key=${API_KEY}`;

export const getSmallPosterImage = imagePath =>
	`${BASE_IMG_PATH}${API.SMALL_IMG_WIDTH}/${imagePath}`;

export const getLargeImage = imagePath => `${BASE_IMG_PATH}${API.LARGE_IMG_WIDTH}/${imagePath}`;

export const getMovieByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieCastByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.CREDITS}`);

export const getPopularMovieEndpoint = () => getAuthorisedEndpoint(API.POPULAR_MOVIES);

export const getRecommendedMoviesByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.RECOMMENDATIONS}`);

export const getSearchMovieEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MOVIES)}&query=${searchTerm}`;

export const getSearchMultiEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MULTI)}&query=${searchTerm}`;

export const getTrendingMoviesByDay = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.DAY}`);

export const getTrendingMoviesByWeek = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.WEEK}`);

export const getTrendingPersonByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.PERSON}${API.WEEK}`);

export const getTrendingTvByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.TV}${API.WEEK}`);

export const getPersonByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}`) + "&append_to_response=images";

export const getPersonCreditsByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}${API.COMBINED_CREDITS}`);

export const getTvByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.TV}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieGenresEndpoint = () => getAuthorisedEndpoint(`${API.GENRES}`);
export const getDiscoverMovieEndpoint = () => getAuthorisedEndpoint(`${API.DISCOVER_MOVIE}`);
