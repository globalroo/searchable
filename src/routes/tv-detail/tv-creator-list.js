import React from "react";

import { getSmallPosterImage } from "src/tmdb-service/tmdb-config";
import { ListItem, ListItemAvatar, Avatar, ListItemText } from "@material-ui/core";
import { MEDIA_TYPE } from "src/tmdb-service/tmdb-api";
import { navigateTo } from "src/helpers/navigate";
import { useImageLoader } from "src/hooks/image-loader";

export const TvCreatorList = ({ id, name, profile_path }) => {
	const [image] = useImageLoader(getSmallPosterImage(profile_path));
	return (
		<ListItem
			key={id}
			alignItems="center"
			className={"pointer-hover"}
			data-testid={`tv-creator-${id}`}
			onClick={() => navigateTo({ media_type: MEDIA_TYPE.PERSON, id })}
		>
			<ListItemAvatar>
				<Avatar style={styles.avatar} alt="name" src={image} />
			</ListItemAvatar>
			<ListItemText primary={name} />
		</ListItem>
	);
};

const styles = {
	avatar: {
		width: 70,
		height: 70
	}
};
