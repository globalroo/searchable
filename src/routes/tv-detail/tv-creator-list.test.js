import React from "react";
import { render, fireEvent } from "react-testing-library";
import { TvCreatorList } from "./tv-creator-list";
import { MEDIA_TYPE } from "src/tmdb-service/tmdb-api";

const mockNavigateSpy = jest.fn();

jest.mock("src/helpers/navigate", () => ({
	navigateTo: val => {
		mockNavigateSpy(val);
	}
}));

const mockCreator = { id: 1, name: `test`, profile_path: `testprofile` };

describe("Selecting a creator avatar routes to their profile", () => {
	it("Renders a tv-creator-list component", () => {
		const { container } = render(<TvCreatorList {...mockCreator} />);
		expect(container.firstChild).toMatchSnapshot();
	});
	it("Renders routes to the correct creator", async () => {
		const { getByTestId } = render(<TvCreatorList {...mockCreator} />);
		const button = getByTestId(`tv-creator-${1}`);
		await fireEvent.click(button);
		expect(mockNavigateSpy).toHaveBeenCalledWith({
			id: mockCreator.id,
			media_type: MEDIA_TYPE.PERSON
		});
	});
});
