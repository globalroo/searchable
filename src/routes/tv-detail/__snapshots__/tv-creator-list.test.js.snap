// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selecting a creator avatar routes to their profile Renders a tv-creator-list component 1`] = `
<li
  class="MuiListItem-root-1 MuiListItem-default-4 MuiListItem-dense-5 MuiListItem-gutters-9 pointer-hover"
  data-testid="tv-creator-1"
>
  <div
    class="MuiAvatar-root-16"
    style="width: 70px; height: 70px;"
  >
    <img
      alt="name"
      class="MuiAvatar-img-18"
      src="http://image.tmdb.org/t/p/w300/testprofile"
    />
  </div>
  <div
    class="MuiListItemText-root-19"
  >
    <span
      class="MuiTypography-root-25 MuiTypography-subheading-32 MuiListItemText-primary-22"
    >
      test
    </span>
  </div>
</li>
`;
